feat: add StacBackdropFilter widget

manav-m opened the pull request:
Adds the StacBackdropFilter widget, which applies a filter to the background of its child widget. The backdrop filter is applied to everything behind the child widget that is within the same layer.

Examples
✅ Example: Basic blur filter
```json
{
  "type": "clipRRect",
  "borderRadius": {
    "all": 16
  },
  "child": {
    "type": "sizedBox",
    "height": 200,
    "width": 300,
    "child": {
      "type": "stack",
      "fit": "expand",
      "children": [
        {
          "type": "container",
          "decoration": {
            "image": {
              "src": "https://images.pexels.com/photos/1323550/pexels-photo-1323550.jpeg",
              "fit": "cover"
            }
          }
        },
        {
          "type": "backdropFilter",
          "filter": {
            "type": "blur",
            "sigmaX": 15.0,
            "sigmaY": 15.0
          },
          "child": {
            "type": "container",
            "decoration": {
              "color": "#80FFFFFF"
            }
          }
        }
      ]
    }
  }
}
```

More examples can be found in the docs at docs/widgets/backdrop_filter.md

Related Issues
Closes [#269](https://github.com/StacDev/stac/issues/269)

Type of Change
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code refactor
- [ ] Build configuration change
- [ ] Documentation
- [ ] Chore

manav-m committed:
feat: add StacBackdropFilter widget

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StacDev/stac?pullRequest=274) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StacDev/stac?pullRequest=274) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StacDev/stac?pullRequest=274) it.</sub>

divyanshub024 commented on packages/stac/lib/src/parsers/widgets/stac_backdrop_filter/stac_backdrop_filter_parser.dart:
```diff
@@ -0,0 +1,37 @@
+      return child;
+    }
+
+    final imageFilter = StacImageFilterParser.parse(model.filter);
+    if (imageFilter == null) {
+      return child;
+    }
+
+    return BackdropFilter(
+      filter: imageFilter,
+      child: child,
+    );
+  }
```
We should throw error when imageFilter is missing. So, ideally we should remove the null check for imageFilter.

divyanshub024 commented on packages/stac/lib/src/parsers/widgets/stac_backdrop_filter/stac_backdrop_filter_parser.dart:
```diff
@@ -0,0 +1,37 @@
+  Widget parse(BuildContext context, StacBackdropFilter model) {
+    final child = Stac.fromJson(model.child, context) ?? const SizedBox();
+
+    if (!model.enabled) {
```
model.enabled should be passed in `BackdropFilter`

divyanshub024 replied:
@manav-m I think this one is not resolved. Can you please check?

divyanshub024 commented on packages/stac/lib/src/parsers/widgets/stac_backdrop_filter/stac_backdrop_filter.dart:
```diff
@@ -0,0 +1,19 @@
+    required StacImageFilter filter,
+    Map<String, dynamic>? child,
+    @Default(true) bool enabled,
+  }) = _StacBackdropFilter;
```
Can we please add the `blendMode` property as well?

divyanshub024 commented on packages/stac/lib/src/parsers/widgets/stac_image_filter/stac_image_filter.dart:
```diff
@@ -0,0 +1,22 @@
+abstract class StacImageFilter with _$StacImageFilter {
+  const factory StacImageFilter({
+    required String type,
+    @Default(StacDouble(10.0)) StacDouble sigmaX,
+    StacDouble? sigmaY,
+    @Default(StacDouble(1.0)) StacDouble radiusX,
+    StacDouble? radiusY,
```
Default value for sigma and radius should be 0.

manav-m replied:
A 0 sigma/radius value would basically make the filter do nothing by default. Shouldn't the filter do a basic blur by default?

divyanshub024 replied:
We usually set the default value of the Flutter framework to keep Stac consistent with Flutter.

divyanshub024 commented on packages/stac/lib/src/parsers/widgets/stac_image_filter/stac_image_filter.dart:
```diff
@@ -0,0 +1,22 @@
+@freezed
+abstract class StacImageFilter with _$StacImageFilter {
+  const factory StacImageFilter({
+    required String type,
```
Can we please turn `type` into an enum?

divyanshub024 commented on packages/stac/lib/src/parsers/widgets/stac_image_filter/stac_image_filter_parser.dart:
```diff
@@ -0,0 +1,48 @@
+import 'package:stac/src/parsers/widgets/stac_image_filter/stac_image_filter.dart';
+
+class StacImageFilterParser {
+  static ImageFilter? parse(StacImageFilter? filter) {
```
Can we please turn this into an extension method?

https://dart.dev/language/extension-methods

divyanshub024 reviewed (changes requested):
Thanks a lot for this contribution @manav-m 🎉 I have few suggestions please check them out.

divyanshub024 committed:
Merge branch 'dev' into feat-backdrop-filter

manav-m committed:
Changes for PR #274
Removed null check for image filter
Passing model.enabled to backdrop filter
Added support for Blendmode Property
Set default value for sigma and radius to 0
Filter type is now an enum
Turn the stac_image_filter_parser into an extension method

manav-m committed:
Remove null checks for model.enabled, imageFilter
Convert the parse function into a getter

divyanshub024 commented on website/docs/widgets/backdrop_filter.md:
```diff
@@ -0,0 +1,231 @@
+|-----------|------------------------|--------------------------------------------------------------------------------------------|
+| `filter`  | `Map<String, dynamic>` | The image filter to apply to the existing painted content before painting the child.       |
+| `child`   | `Map<String, dynamic>?`| The widget to paint after applying the filter to the existing painted content.             |
+| `enabled` | `bool`                 | Whether the filter should be applied. When false, the child is drawn without a filter. Defaults to `true`. |
```
Can we please add the `blendMode` in docs?

manav-m committed:
Add Blendmode property to docs

divyanshub024 reviewed (approved):
Thanks for your contribution @manav-m 🎉 💯
